### DISCUSSION TITLE
How to bulk update table on Postgres using ORM

### DISCUSSION THREAD

**Discussion by rst1** [sqlalchemy/sqlalchemy#10014] [Usage Questions] [answered]
Hello
I have a case where sometimes I want to add rows to a table and sometimes I want to update a small number of fields in the same table.
I am working with a total dataset on the order of 100 million rows, and I am adding or editing batches on the order of 50,000 rows, so I can't do much manual adjustment of the data. 
If possible, I want the Update and Insert statements to use data in the same format

I am writing to PostgreSQL 15, but I want my solution to be backwards compatible to PostgreSQL 12

I have the Insert part working, but I can't figure out how to do the Update.
I am fine with using "upsert" or update, or another technique, if there is one.
I think I almost have the upsert working, except I can't figure out what to put in the blank "________ " that is indicated below, but I am fine with using a different method, if there is a better way to do it.

A cut down version of my code along with a simplified version of the data is included below.
The table definition is also included,
Note that the insert and the update both happen on the same table.
the update uses a subset of the data used by the insert (that is, fewer columns.)
I want the update to update any data that is available to it, and leave all other columns unchanged. 

```from sqlalchemy import MetaData, Table, create_engine, Column
from sqlalchemy.orm import Session as db_Session
from sqlalchemy.dialects import postgresql

#CREATE TABLE IF NOT EXISTS public.request_info
#(
#    id uuid NOT NULL,
#    process_timestamp timestamp without time zone NOT NULL,
#    workspace character varying(127) COLLATE pg_catalog."default" NOT NULL,
#    user_id integer NOT NULL,
#    request_method character varying(15) COLLATE pg_catalog."default" NOT NULL,
#    result_code integer NOT NULL,
#    environment character varying(500) COLLATE pg_catalog."default",
#    CONSTRAINT request_info_pkey PRIMARY KEY (id)
#);


mode  = 'fixDates' #alternative is #mode  = 'addRows'
if mode == 'addRows':
	data = [{
        'id': 'eeec90f8-0b9d-49af-8c2d-c7e76756f6b0',
        'process_timestamp': '2023-06-21 17:00:00.088+00:00',
        'workspace': 'W1',
        'user_id': 78,
        'request_method': 'GET',
        'result_code': 200,
        'environment': 'E1'
		}, 
		{
        'id': '48081535-470b-4a36-ac31-a7b40bb45118',
        'process_timestamp': '2023-06-21 17:00:00.179+00:00',
        'workspace': 'W2',
        'user_id': 116,
        'request_method': 'GET',
        'result_code': 200,
        'environment': 'E2'
		}]
else:
	data = [{
        'id': 'eeec90f8-0b9d-49af-8c2d-c7e76756f6b0',
        'process_timestamp': '2023-06-21 17:00:00.088+00:00',
        'environment': 'E1'
		}, 
		{
        'id': '48081535-470b-4a36-ac31-a7b40bb45118',
        'process_timestamp': '2023-06-21 17:00:00.179+00:00',
        'environment': 'E2'
		}]

engine = create_engine("postgresql+psycopg://postgres:<password>@localhost:5432/request_info")
engine = engine.execution_options(isolation_level="AUTOCOMMIT")
metadata_obj = MetaData()
metadata_obj.reflect(bind=engine)
sess=db_Session(engine)

insert_table = Table('request_info', metadata_obj, autoload_with=engine)
    
pg_insert = postgresql.insert(insert_table).returning(Column('id'))
pg_insert = pg_insert.on_conflict_do_nothing(index_elements=['id'])

pg_update = postgresql.insert(insert_table).returning(Column('id'))
pg_update = pg_update.on_conflict_do_update(constraint='id', set_ = _____________)


    if mode == 'addRows':
        result=sess.execute( pg_insert,  data)
        sess.commit()
    else:
        result=sess.execute( pg_update,  data)
        sess.commit()
sess.close()

**Comment by CaselIT** [comment]
Hi,

Have you looked in the docs? It illustrates various options https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#insert-on-conflict-upsert

**Reply by rst1** [reply]
I spent quite a bit of time in the docs, a lot of it on that page, but couldn't figure out what I was missing. 
(Maybe it's me, but I find these docs hard to translate into code.)

I also spent quite a while on Google, but again, I didn't find anything that seemed applicable to my situation.

(Responding with RTFM is fair, since I didn't actually say I already did that. :-) )

**Reply by CaselIT** [reply, marked as answer]
So here I'm assuming you want to update the existing row with what you have in the data.

In that case `set_` should be a dict where keys ate column names to set and values are taken from `insert_stmt.excluded.`
Something like this
```py
stmt= insert()
stmt = stmt.on_conflict_do_update(
  ..., set_={'col1':stmt.excluded.col1, ...}
)

**Reply by rst1** [reply]
Thank you, this was the piece that I was missing!

**Comment by rst1** [comment]
I'll provide a few more details about what I tried.
set_ = 
Should have either a dict or a columnCollection 

The data that is being provided to the insert is a list of dictionaries, 
- I don't know how to provide the correct dictionary.
- I thought a columnCollection made up of the columns that I wanted updated might work. I'm not sure how to create a columnCollection with just a few columns. Nor am I sure if that would work.

**Comment by gordthompson** [comment]
One possibility would be to pull the new data into a pandas DataFrame and then use this method to do the upsert

https://gist.github.com/gordthompson/ae7a1528fde1c00c03fdbb5c53c8f90f
